### PR TITLE
API: membership checks on ExtensionArray containing NA values

### PR DESCRIPTION
**Pull request by topper-123**
Membership checks on ExtensionArrays containing `NA` values raises ValueError in some circumstances (but not in other):

```python
>>> arr1 = pd.array(["a", pd.NA])
>>> arr2 = pd.array([pd.NA, "a"])
>>> "a" in arr1
True  # ok
>>> "a" in arr2
TypeError: boolean value of NA is ambiguous  # not ok
>>> pd.NA in arr1
TypeError: boolean value of NA is ambiguous  # not ok
>>> pd.NA in arr2
True  # ok
```

So overall quite random failures. This PR fixes this problem by adding a custom `__contains__` method on `ExtensionArray`.

I assume that we want `pd.NA in arr1` to keep returning True. Note however that `np.nan in np.array([np.nan])` return False, so pandas' behaviour is different.

### CODE REVIEW

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -351,6 +351,17 @@ def __iter__(self):
...
             yield self[i]
 
+    def __contains__(self, item) -> bool:
+        """
+        Return for `item in self`.
+        """
+        # comparisons of any item to pd.NA always return pd.NA, so e.g. "a" in [pd.NA]
+        # would raise a TypeError. The implementation below works around that.
+        if isna(item):
+            return False
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:361]: is this going to affect e.g. DatetimeArray that does contain a NaT?

Comment (topper-123) [reply pandas/core/arrays/base.py:361]: I made a mistake in OP. Can you please read it again. `NaT in arr` should apparantæly return True, if an Nat is in the array.

I'll update the PR.

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -351,6 +351,17 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def __contains__(self, item) -> bool:
+        """
+        Return for `item in self`.
+        """
+        # comparisons of any item to pd.NA always return pd.NA, so e.g. "a" in [pd.NA]
+        # would raise a TypeError. The implementation below works around that.
+        if isna(item):
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:360]: is_valid_nat_for_dtype

Comment (topper-123) [reply pandas/core/arrays/base.py:360]: I don't understand you here. Are you saying that `pd.NaT in arr` should return False if the array is not a datetime-like array? Can you expand a bit?

Comment (jbrockmendel) [reply pandas/core/arrays/base.py:360]: > Are you saying that pd.NaT in arr should return False if the array is not a datetime-like array? 

correct (unless maybe object dtype).  Some other examples:

```
dta = pd.date_range("2016-01-01", periods=3)._data  # <-- DateTimeArray
dta[-1] = pd.NaT
pa = dta.to_period("D")  # <-- PeriodArray
tda = dta - dta  # <-- TimedeltaArray
arr = pd.array([1, 2 ,3, pd.NA])  # <-- IntegerArray
flt = pd.array([1.0, 2.0, 3.0, np.nan])  # <-- FloatArray

tdnat = np.timedelta64("NaT", "ns")
dtnat = np.timedelta64("NaT", "ns")

>>> tdnat in dta
False
>>> dtnat in dta
True  # <-- actually this returns False, but thats a newly-discovered bug

>>> tdnat in pa
False
>>> dtnat in pa
False

>>> tdnat in tda
True   # <-- actually this raises TypeError, but thats a newly-discovered bug
>>> dtnat in tda
False    # <-- actually this raises TypeError, but thats a newly-discovered bug

>>> tdnat in flt
False
>>> dtnat in flt
False    # <-- actually this raises TypeError, but thats a newly-discovered bug

>>> tdnat in arr
False
>>> dtnat in arr
False     # <-- actually this raises TypeError, but thats a newly-discovered bug
```

side-note after checking all of these: yikes!

**Comment by jreback** [comment]
ok so this looks fine. merge and then open issues / followup for recently discovered cases?

File: pandas/core/arrays/base.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/arrays/base.py:360]: I don't use datetime-likes very much, so I'm a bit weak on what these nan-like comparisons to nan-likes should return in datatime-likes and you certainly didn't help there :-)

Maybe do as @jreback suggests, and handle these later/separately? An alternative could be in the base `Expensionarray` check for `pd.NA` only. It sub-classes wan to do it differently, they can implement a `__contains__` method themselves. That seems the cleanest to me.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
I think we should first discuss / decide on what behaviour we actually want for containment involving missing values (as the item being checked). 

As it seems (even apart from the noted errors with pd.NA) an inconsistent situation right now? (np.nan gives False, pd.NaT gives True)

Also, can you add a base extension test?

**Comment by jorisvandenbossche** [comment]
Now, I don't really know myself at the moment what I would expect for behaviour here. Listing all possible options (eg for `pd.NA in pd.array([1, 2, pd.NA])`:

* Return True (because pd.NA is present in the array)
* Return False (because pd.NA is not equal to itself, so cannot be "found")
* Raise an error (because it is ambiguous (since there is an argument for both True/False), or because there is simply no "correct" answer, in the same logic as we also raise an error for `bool(pd.NA)`)
  * Counterargument: in dictionaries pd.NA can be used as key, so you could argue that containement is similar as dictionary key checking
* Return pd.NA (similar reasons as above, + since the pd.NA values are considered as "unknown", it is *also* "unknown" if the unknown value is contained in the array)

Quickly looking at some other languages to see what they do:

R returns True:

```R
> NA %in% c(1, 2, 3, NA)
[1] TRUE
```

Julia is "strict" about the unknown part of `missing`, and if a missing value is present in the array, even any other value not present in the array is not known to *not* be in the array (so not even returns False):

```julia
julia> arr = [1 2 3 missing]
1×4 Array{Union{Missing, Int64},2}:
 1  2  3  missing

julia> 1 in arr
true

julia> 5 in arr
missing

julia> missing in arr
missing

julia> arr = [1 2 3]
1×3 Array{Int64,2}:
 1  2  3

julia> 5 in arr
false

julia> missing in arr
missing
```

**Comment by topper-123** [comment]
Yes, @jorisvandenbossche, my doubts on nan-likes as well: There doesn't seem to be any inherantly correct choice, as it look to me. 

We have however already in pandas that:

```python
>>> arr2 = pd.array([pd.NA, "a"])
>>> pd.NA in arr2  # see OP
True
>>> dta = pd.date_range("2016-01-01", periods=2)._data
>>> dta[0] = pd.NaT
>>> pd.NaT in dta
True
```

So from at backward compat stand point there is some argument for returning True in Pandas, if looking for a nan-like in the ExtensionArray.

**Comment by jorisvandenbossche** [comment]
Specifically that `pd.NA in pd.array([pd.NA, "a"])` is returning True, I wouldn't use as a "prior" art argument, as I think this is mostly accidental behaviour and we never really considered yet the desired behaviour for `__contains__` for nullable EAs (and those types are not yet considered stable, so we can change behaviour if we decide this is needed).
 
The fact that it returns True for pd.NaT certainly has a much longer history, but I don't know to what extent this was intentional (@jbrockmendel ?). For example the numpy counterpart actually returns False:

```
In [20]: arr = np.array(["NaT"], dtype="datetime64[ns]")

In [21]: arr[0]
Out[21]: numpy.datetime64('NaT')

In [22]: arr[0] in arr
Out[22]: False
```

---

Something else, the current `isna(item)` check is also too broad, as (I think from looking at the code), it will make that things like `None in pd.array([pd.NA, "a"])` or `np.nan in pd.array([pd.NA, "a"])` will also return True, which I think we certainly don't want?

**Comment by jbrockmendel** [comment]
> The fact that it returns True for pd.NaT certainly has a much longer history, but I don't know to what extent this was intentional

I'm pretty sure that is unintentional.  I've been fixing some `isna` checks to use `is_valid_nat_for_dtype`, but haven't gotten all of them.

> Something else, the current `isna(item)` check is also too broad, as (I think from looking at the code), it will make that things like `None in pd.array([pd.NA, "a"])` or `np.nan in pd.array([pd.NA, "a"])` will also return `True`, which I think we certainly don't want?

I'd advocate updating `is_valid_nat_for_dtype` to reflect those exclusions.  (And renaming `valid_nat` -> `valid_na`)

> ```
> In [22]: arr[0] in arr
> Out[22]: False
> ```

That is surprising.  My intuition is that `arr[0] in arr` should be `True` for _any_ `arr`.  I'm open to being convinced otherwise.

**Comment by jorisvandenbossche** [comment]
> That is surprising.  My intuition is that `arr[0] in arr` should be `True` for _any_ `arr`.  I'm open to being convinced otherwise.

For NaN/NaT, those values are not equal to itself, and I think numpy uses equality to implement `__contains__`

**Comment by jbrockmendel** [comment]
> For NaN/NaT, those values are not equal to itself, and I think numpy uses equality to implement __contains__

My intuition is based more around our Index implementations that use `get_loc`.

What would the idiomatic way of checking `any(x is pd.NA for x in arr)` be?

**Comment by topper-123** [comment]
> What would the idiomatic way of checking `any(x is pd.NA for x in arr)` be?

`arr.isna().any()` IMO.

I think there will be be a surprise if users check for nan-likes using `__contains__` for some users, bescause there are two reasoably choices:

1. `pd.NA in arr` should be interpreted as `pd.NA == arr).any()`, resulting in `False` even if a nan-like is in `arr`
2. `pd.NA in arr` should be interpreted as `arr.isna().any()`, resulting in `True` if a nan-like is in `arr`, but will also return True for e.g. `pd.NaT in arr`, which may be unintended (or has to be decided upon.

A solution to avoid ambiguity is to raise a TypeError, if a nan-like is supplied to `__contains__`: 

```python
    def __contains__(self, item) -> bool:
        if isna(item):
            raise TypeError("NA value not allowed. Use .isna method to check for NA values.")
        else:
            return (item == self).any()
```

This would at least ensure that `"a" in arr` doesn't raise unexpectedly.

**Comment by jbrockmendel** [comment]
>> What would the idiomatic way of checking any(x is pd.NA for x in arr) be?

> arr.isna().any() IMO.

My complaint about that is it doesn't distinguish between different NAs, most relevant for object dtype.

> This would at least ensure that "a" in arr doesn't raise unexpectedly.

Short-term, can we fix this without taking a stance on the NA topic?

My intuition is that `obj in pd.array(arr)` should match `obj in pd.Index(arr)` and `obj in pd.Series(arr)`.  Is this controversial?

**Comment by jreback** [comment]
> > > What would the idiomatic way of checking any(x is pd.NA for x in arr) be?
> 
> > arr.isna().any() IMO.
> 
> My complaint about that is it doesn't distinguish between different NAs, most relevant for object dtype.
> 
> > This would at least ensure that "a" in arr doesn't raise unexpectedly.
> 
> Short-term, can we fix this without taking a stance on the NA topic?
> 
> My intuition is that `obj in pd.array(arr)` should match `obj in pd.Index(arr)` and `obj in pd.Series(arr)`. Is this controversial?

agree with @jbrockmendel 

I think have NaN != itself is different then a membership test. we do not want special cases.

File: pandas/tests/arrays/categorical/test_operators.py [hunk 3]
```diff
@@ -395,3 +395,20 @@ def test_numeric_like_ops(self):
...
+        # GH-37867
+        cat = Categorical(["a", "b"], ordered=ordered)
+        assert "a" in cat
+        assert "x" not in cat
+        assert pd.NA not in cat
+
+        cat = Categorical([np.nan, "a"], ordered=ordered)
+        assert "a" in cat
+        assert "x" not in cat
+        assert pd.NA in cat
```
Comment (jreback) [inline pandas/tests/arrays/categorical/test_operators.py:409]: can you tests `np.nan` here as well (True), while `pd.NaT` will be False

File: pandas/tests/arrays/string_/test_string.py [hunk 4]
```diff
@@ -345,3 +345,22 @@ def test_astype_from_float_dtype(dtype):
...
+    # GH-37867
+    arr = pd.arrays.StringArray(np.array(["a", "b"], dtype=object))
+
+    assert "a" in arr
+    assert "x" not in arr
+    assert pd.NA not in arr
+
+    arr = pd.arrays.StringArray(np.array(["a", pd.NA]))
+    assert "a" in arr
+    assert "x" not in arr
```
Comment (jreback) [inline pandas/tests/arrays/string_/test_string.py:360]: np.nan should be True here?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
> Short-term, can we fix this without taking a stance on the NA topic?

How is fixing this not taking a stance? ;) 
Because to fix it we need to choose a certain behaviour regardless, which then defacto becomes the `__contains__` behaviour for pd.NA

The `obj in pd.Index(arr)` example from @jbrockmendel is a good comparison, because it basically does a engine/hashtable lookup operation under the hood and so I suppose will be consistent with indexing (`series.loc[pd.NA]` if the index has missing values). 

So that basically gives two possible interpretations / implementations of `__contains__`:

- equality-based: (`(arr == obj).any()`)
- lookup-based (eg `pd.NA in dict.fromkeys(arr)`, or our indexing engine lookup)

While numpy uses the first one, it seems to make sense to use the second option for membership test like `__contains__`.  

But, I think there is also something to say for using the "unknown value" interpretation of `pd.NA`. And then we don't know if some "unknown value" is present in the array, and we shouldn't return True or False.

File: pandas/tests/arrays/categorical/test_operators.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/categorical/test_operators.py:409]: Shouldn't `pd.NA in cat` not give False? Since it is np.nan that Categorical is using as missing value indicator, and not pd.NA

File: pandas/tests/arrays/string_/test_string.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/string_/test_string.py:360]: or False? 
Should we distinguish between np.nan vs pd.NA ? (currently the indexing engines do distinguish both)

File: pandas/tests/arrays/string_/test_string.py [hunk 5]
```diff
@@ -345,3 +345,22 @@ def test_astype_from_float_dtype(dtype):
     result = s.astype("string")
     expected = pd.Series(["0.1"], dtype="string")
     tm.assert_series_equal(result, expected)
+
+
+def test_contains():
+    # GH-37867
+    arr = pd.arrays.StringArray(np.array(["a", "b"], dtype=object))
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/string_/test_string.py:352]:
```suggestion
    arr = pd.array(["a", "b"], dtype="string")
```

is a bit shorter to construct those

File: pandas/tests/arrays/categorical/test_operators.py (hunk 3 above)
Comment (topper-123) [reply pandas/tests/arrays/categorical/test_operators.py:409]: In the current implementation, they all give True, so `pd.NaT in cat` gives True also, This is because `cat` is object-dtype, so can contains a `pd.NaT`.

File: pandas/tests/arrays/string_/test_string.py (hunk 4 above)
Comment (topper-123) [reply pandas/tests/arrays/string_/test_string.py:360]: They all (np.nan, pd.NA and pd.NaT) True here. The reason is that:

```python
>>> pd.core.dtypes.missing.is_valid_nat_for_dtype(pd.NaT, pd.StringDtype())
True
```

This looks like a different bug than this PR.

Comment (jorisvandenbossche) [reply pandas/tests/arrays/string_/test_string.py:540]: It's a direct consequence of the implementation, and I think actually easy to fix. Eg maybe you can use `if val is self.dtype.na_value` instead of `isna(val)` or `is_valid_nat_for_dtype(val, ..)` ?

**Comment by jorisvandenbossche** [comment]
@topper-123 since you are adding a method on the base extension array class, I think we should also add a test for this in the generic base extension tests. There are fixtures for data with/without missing values and for the na_value, so it should be possible to test this in general.

**Comment by topper-123** [comment]
> 
> 
> @topper-123 since you are adding a method on the base extension array class, I think we should also add a test for this in the generic base extension tests. There are fixtures for data with/without missing values and for the na_value, so it should be possible to test this in general.

I've looked, but can't find the base extension tests. Could you say where they're located?

**Comment by jorisvandenbossche** [comment]
In `pandas/tests/extension/base/`, and then in one of those files. Not sure which file is best fitting, maybe `interface.py`, or `methods.py`

**Comment by jbrockmendel** [comment]
>> Short-term, can we fix this without taking a stance on the NA topic?

> How is fixing this not taking a stance? ;)

I was specifically referring to "This would at least ensure that `"a" in arr` doesn't raise unexpectedly.", which I think we can ensure without taking a stance on the behavior of `pd.NA in arr`.

**Comment by jorisvandenbossche** [comment]
> I was specifically referring to "This would at least ensure that `"a" in arr` doesn't raise unexpectedly."

Ah, yes, that's something we certainly agree on I think that this at least shouldn't raise! 
(but to fix that we still need to choose a behaviour for `pd.NA in ..` as well)

Now, I am fine with following our behaviour for Index containment / indexing, which is to treat NaN and pd.NA etc equal to itself (in contrast to `==` equality).

**Comment by topper-123** [comment]
I've updated the PR.

It nows returns True if the na value is `arr.dtype.na_value` else False.

**Comment by topper-123** [comment]
The Travis failure looks unrelated. I'll run the PR again after comments.

File: pandas/core/arrays/base.py [hunk 6]
```diff
@@ -354,6 +355,19 @@ def __iter__(self):
...
+    def __contains__(self, item) -> bool:
+        """
+        Return for `item in self`.
+        """
+        # comparisons of any item to pd.NA always return pd.NA, so e.g. "a" in [pd.NA]
+        # would raise a TypeError. The implementation below works around that.
+        if item is self.dtype.na_value:
+            return isna(self).any() if self._can_hold_na else False
+        elif is_scalar(item) and isna(item):
+            return False
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/base.py:367]: This could maybe be done in a separate PR, as it is FloatingArray specific (and the NaN behaviour is not yet fully fleshed out), but this check will do the wrong thing for `np.nan` if the FloatingArray actually contains a NaN (as `self.dtype._na_value` is then pd.NA and not np.nan).

File: pandas/tests/extension/base/interface.py [hunk 7]
```diff
@@ -29,6 +29,24 @@ def test_can_hold_na_valid(self, data):
...
+    def test_contains(self, data):
+        # GH-37867
+        scalar = data[~data.isna()][0]
+
+        assert scalar in data
+
+        na_value = data.dtype.na_value
+
+        if data.isna().any():
+            assert na_value in data
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/interface.py:41]: `data` is not guaranteed to have a missing values, but there is also a `data_missing` fixture that has this guarantee and which you can add here to the test, and use that here. 
And then to be sure we test the case of data without NAs, maybe do the else with `data[~data.isna()]` (that is guaranteed to not be empty)

Comment (jorisvandenbossche) [reply pandas/tests/extension/base/interface.py:41]: And thanks for this base test!

File: pandas/core/arrays/base.py (hunk 6 above)
Comment (topper-123) [reply pandas/core/arrays/base.py:367]: So a `FloatingArray` can have multiple nan types? Is that really necessary?

I think I would prefer that would be handled in `FloatingArray`, so I don't have to stretch this PR too far.

File: pandas/tests/extension/base/interface.py (hunk 7 above)
Comment (topper-123) [reply pandas/tests/extension/base/interface.py:41]: Ok, I've added a test method using data_missing.

File: pandas/core/arrays/base.py [hunk 8]
```diff
@@ -354,6 +355,19 @@ def __iter__(self):
...
             yield self[i]
 
+    def __contains__(self, item) -> bool:
+        """
+        Return for `item in self`.
+        """
+        # comparisons of any item to pd.NA always return pd.NA, so e.g. "a" in [pd.NA]
+        # would raise a TypeError. The implementation below works around that.
+        if item is self.dtype.na_value:
+            return isna(self).any() if self._can_hold_na else False
```
Comment (jreback) [inline pandas/core/arrays/base.py:365]: can you use `self.isna()`

File: pandas/core/arrays/base.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:367]: > So a FloatingArray can have multiple nan types? Is that really necessary?

Not multiple "nan types", to be nitpicky, since pd.NA is not a "nan". But yes, it can contains both pd.NA and np.nan at the moment. But that's not yet fully decided though, long discussion at https://github.com/pandas-dev/pandas/issues/32265/. Happy to hear your thoughts about it.

> I think I would prefer that would be handled in FloatingArray, so I don't have to stretch this PR too far.

Yes, will put it on my list of follow to-do items for FloatingArray that I need to open. So don't worry about it here

File: pandas/tests/extension/base/interface.py [hunk 9]
```diff
@@ -29,6 +29,43 @@ def test_can_hold_na_valid(self, data):
         # GH-20761
         assert data._can_hold_na is True
 
+    def test_contains(self, data):
+        # GH-37867
```
Comment (jreback) [inline pandas/tests/extension/base/interface.py:33]: can you put a brief comment here on this discussion we are having (e.g. to distinguuish between different nans and so on ), for a future reader

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm, some comments.

File: pandas/tests/extension/base/interface.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/interface.py:60]: BTW, you can pass both fixtures to the same test, then you can write the test with less duplication as it is now

Comment (jreback) [reply pandas/tests/extension/base/interface.py]: you can pass them as the string name of the test and use: `frame = request.getfixturevalue(fixture_func_name)` to create the actual frame

File: pandas/core/arrays/base.py [hunk 10]
```diff
@@ -354,6 +355,19 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def __contains__(self, item) -> bool:
+        """
+        Return for `item in self`.
+        """
+        # comparisons of any item to pd.NA always return pd.NA, so e.g. "a" in [pd.NA]
```
Comment (jreback) [inline pandas/core/arrays/base.py:362]: can you add this issue number here for a future refernce

File: pandas/tests/extension/base/interface.py [hunk 11]
```diff
@@ -29,6 +29,31 @@ def test_can_hold_na_valid(self, data):
         # GH-20761
         assert data._can_hold_na is True
 
+    def test_contains(self, data, data_missing):
+        # GH-37867
+        # Tests for membership checks. Membership checks for nan-likes is tricky and
+        # the settled on rule is: `nan_like in arr` is True if nan_like is
+        # arr.dtype.na_value and arr.isna().any() is True. Else the check returns False.
+
+        for this_data in [data, data_missing]:
```
Comment (jreback) [inline pandas/tests/extension/base/interface.py:38]: see my comments below on how to actually parameterize this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py (hunk 10 above)
Comment (topper-123) [reply pandas/core/arrays/base.py:362]: Ok

File: pandas/tests/extension/base/interface.py (hunk 11 above)
Comment (topper-123) [reply pandas/tests/extension/base/interface.py:38]: I don't its possible to parametrize fixtures? do have an example or hint on how that's done?

Comment (jreback) [reply pandas/tests/extension/base/interface.py:38]: grep for

pytest.getfixturevalue in tests
we just had a use of this

Comment (topper-123) [reply pandas/tests/extension/base/interface.py:38]: Couldn't get it to work, I'll try it again tonight.

Comment (jorisvandenbossche) [reply pandas/tests/extension/base/interface.py:38]: Sorry, but I would simply leave it as is, using `pytest.getfixturevalue ` only makes it way more complicated as it needs to be

File: pandas/tests/extension/base/interface.py [hunk 12]
```diff
@@ -29,6 +29,30 @@ def test_can_hold_na_valid(self, data):
...
+            if this_data.isna().any():
+                assert na_value in this_data
+            else:
+                assert na_value not in this_data
+
+            # this_data can never contain other nan-likes than na_value
+            for na_value_type in {None, np.nan, pd.NA, pd.NaT}:
+                if na_value_type is na_value:
+                    continue
+                assert na_value_type not in this_data
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/interface.py:54]: Could also write it like this without a for-loop (and with ensuring we are actually testing with one array without missing values, and without the if isna().any() check), which I find easier to follow (but that's probably subjective ;))

```suggestion
        na_value = data.dtype.na_value
        # ensure data without missing values
        data = data[~data.isna()]

        # first elements are non-missing
        assert data[0] in data
        assert data_missing[0] in data_missing

        # check the presence of na_value
        assert na_value in data_missing
        assert na_value not in data

        # the data can never contain other nan-likes than na_value
        for na_value_type in {None, np.nan, pd.NA, pd.NaT}:
            if na_value_type is na_value:
                continue
            assert na_value_type not in data
            assert na_value_type not in data_missing
```

Comment (topper-123) [reply pandas/tests/extension/base/interface.py]: Yeah, I like that better too.

File: pandas/tests/extension/arrow/test_bool.py [hunk 13]
```diff
@@ -50,6 +50,10 @@ def test_view(self, data):
         # __setitem__ does not work, so we only have a smoke-test
         data.view()
 
+    @pytest.mark.xfail(raises=AssertionError, reason="Not implemented yet")
+    def test_contains(self, data, data_missing):
+        super().test_contains(data, data_missing)
```
Comment (topper-123) [inline pandas/tests/extension/arrow/test_bool.py:55]: The Arrow arrays have Arrow Nulls as the dtype.na_value, while data_missing[0] gives ´None´.

Maybe @TomAugspurger could look into that as part of his Arrow work?

**Comment by topper-123** [comment]
The travis failure looks unrelated and I can't reproduce it on my computer. So IMo this PR passes currently, but I could make another run, if there's agreement on the content.

**Comment by jorisvandenbossche** [comment]
Yes, don't worry about the travis failure on this PR

File: pandas/tests/extension/arrow/test_bool.py (hunk 13 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/arrow/test_bool.py:55]: > The Arrow arrays have Arrow Nulls as the dtype.na_value, while data_missing[0] gives ´None´.

It's fine to simply skip it as you did. This are only test EAs, that were needed for certain specific tests, so no problem this don't fully work otherwise. 

The actual public arrays using Arrow under the hood (eg the new ArrowStringArray) has a different implementation and is fully tested.

File: pandas/core/arrays/categorical.py [hunk 14]
```diff
@@ -1745,6 +1745,8 @@ def __contains__(self, key) -> bool:
         """
         Returns True if `key` is in this Categorical.
         """
+        # in pandas 2.0, remove this method.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/categorical.py:1748]: What does this exactly mean?

File: pandas/tests/extension/test_categorical.py [hunk 15]
```diff
@@ -87,6 +87,31 @@ def test_memory_usage(self, data):
         # Is this deliberate?
         super().test_memory_usage(data)
 
+    def test_contains(self, data, data_missing):
+        # GH-37867
+        # na value handling in Categorical.__contains__ is deprecated.
+        # See base.BaseInterFaceTests.test_contains for more details.
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_categorical.py:93]: This duplicates what you have in `tests/arrays/categorical/test_operators.py ` ?

Comment (topper-123) [reply pandas/tests/extension/test_categorical.py:93]: Unfortunately not: `Categorical` already had a `__contains__` method and it's more permissive than the new one. So, here we have below `assert na_value_type in data_missing`, while the base tests method is `assert na_value_type not in data_missing` (notice the `not`).

na values is also more complicated in categoricals, because in some cases we want to accept `pd.NaT` and in other cases not. I'd like to take it in another round (or let it slide)

Comment (topper-123) [reply pandas/tests/extension/test_categorical.py:93]: This was a mistake, I will remove it.

File: pandas/core/arrays/categorical.py (hunk 14 above)
Comment (topper-123) [reply pandas/core/arrays/categorical.py:1748]: This was a mistake, I will remove it.

File: pandas/core/arrays/numpy_.py [hunk 16]
```diff
@@ -51,6 +51,13 @@ def numpy_dtype(self) -> np.dtype:
         """
         return self._dtype
 
+    @property
+    def na_value(self) -> object:
+        if issubclass(self.type, np.floating):
```
Comment (jreback) [inline pandas/core/arrays/numpy_.py:56]: isn't this *always* nan?

File: pandas/tests/extension/decimal/array.py [hunk 17]
```diff
@@ -155,6 +155,14 @@ def __setitem__(self, key, value):
     def __len__(self) -> int:
         return len(self._data)
 
+    def __contains__(self, item) -> bool:
+        if not isinstance(item, decimal.Decimal):
```
Comment (jreback) [inline pandas/tests/extension/decimal/array.py:159]: does this have actual tests?

File: pandas/tests/extension/json/test_json.py [hunk 18]
```diff
@@ -143,6 +143,11 @@ def test_custom_asserts(self):
         with pytest.raises(AssertionError, match=msg):
             self.assert_frame_equal(a.to_frame(), b.to_frame())
 
+    @pytest.mark.xfail(reason="comparison method not implemented on JSONArray")
```
Comment (jreback) [inline pandas/tests/extension/json/test_json.py:146]: is than issue number for this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/extension/json/test_json.py (hunk 18 above)
Comment (topper-123) [reply pandas/tests/extension/json/test_json.py:146]: The issue number is the first line in the method?

File: pandas/tests/extension/decimal/array.py (hunk 17 above)
Comment (topper-123) [reply pandas/tests/extension/decimal/array.py:159]: Yes, in `pandas/tests/extension/decimal/test_decimal.py` there is `TestInterface`, which inherits from `BaseInterfaceTests`.

File: pandas/tests/extension/json/test_json.py (hunk 18 above)
Comment (jreback) [reply pandas/tests/extension/json/test_json.py:146]: in the xfail pls

File: pandas/core/arrays/numpy_.py (hunk 16 above)
Comment (topper-123) [reply pandas/core/arrays/numpy_.py:56]: The problem was that `np.float64("nan") is not np.nan`. However, I've fixed this, so this method is actuallt not necessary.

File: pandas/tests/extension/json/test_json.py (hunk 18 above)
Comment (topper-123) [reply pandas/tests/extension/json/test_json.py]: ok.

File: pandas/tests/extension/test_categorical.py (hunk 15 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/test_categorical.py:93]: To be clear, I was not referring to the base tests that this is overriding, but the original test you added to `tests/arrays/categorical/test_operators.py ` which also tests this more permissive behaviour?

Comment (topper-123) [reply pandas/tests/extension/test_categorical.py:93]: Yeah. I'll delete the ones in `tests/arrays/categorical/test_operators.py`.

File: pandas/tests/extension/base/interface.py [hunk 19]
```diff
@@ -29,6 +29,31 @@ def test_can_hold_na_valid(self, data):
...
+
+        # first elements are non-missing
+        assert data[0] in data
+        assert data_missing[0] in data_missing
+
+        # check the presence of na_value
+        assert na_value in data_missing
+        assert na_value not in data
+
+        # the data can never contain other nan-likes than na_value
```
Comment (jreback) [inline pandas/tests/extension/base/interface.py:50]: might consider actually using the nulls_fixture as this is more comprehensive (sure it will run the other checks multiple times but no big deal).

File: pandas/tests/extension/test_categorical.py [hunk 20]
```diff
@@ -87,6 +87,31 @@ def test_memory_usage(self, data):
...
+
+        # first elements are non-missing
+        assert data[0] in data
+        assert data_missing[0] in data_missing
+
+        # check the presence of na_value
+        assert na_value in data_missing
+        assert na_value not in data
+
+        # the data can never contain other nan-likes than na_value
```
Comment (jreback) [inline pandas/tests/extension/test_categorical.py:107]: same comment here.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good and fine merging like this, but should change to use the nulls_fixture which is more comprehensive. if you can do this today can get this in

File: pandas/tests/extension/base/interface.py (hunk 19 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/interface.py:50]: I am personally -1 on using a fixture for this (we could move the list of nulls that is currently used for defining the fixture to constant in `_testing.py` and use that constant in both places, though)

Comment (jreback) [reply pandas/tests/extension/base/interface.py:50]: in this case?

we already have a nulls fixture and use it in a great many places

we need to be comprehensive and general in testing - specific cases are ok sometimes

Comment (jreback) [reply pandas/tests/extension/base/interface.py:50]: ok here's the fixture contents
```
@pytest.fixture(params=[None, np.nan, pd.NaT, float("nan"), pd.NA], ids=str)
```

so if you add `float('nan')` then this should cover

Comment (topper-123) [reply pandas/tests/extension/base/interface.py]: I like the generality of using the nulls_fixture too. I've added it in the newest commit.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. thanks @topper-123

File: pandas/tests/extension/arrow/test_bool.py (hunk 13 above)
Comment (xhochy) [reply pandas/tests/extension/arrow/test_bool.py]: Should we also base these test arrays on `pd.NA`?  Otherwise we can use the same code as in the string array to replace the scalar return value of `None` with `na_value`.
